### PR TITLE
Add two examples for Keras: ResNet for cifar10, DisturbLabel

### PR DESCRIPTION
**Pull request by pengpaiSH**
@fchollet 

Add two examples for Keras: ResNet for cifar10, DisturbLabel.

resnet_cifar10.py reproduces paper "Deep Residual Learning for Image Recognition".

disturb_label.py reproduces paper "DisturbLabel: Regularizing CNN on the Loss Layer."

### CODE REVIEW

File: examples/resnet_cifar10.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+from keras.layers import Input, merge
+from keras.models import Model
+from keras.utils import np_utils
+from keras.callbacks import ModelCheckpoint
+from keras.datasets import cifar10
+from keras.preprocessing.image import ImageDataGenerator
+from keras.regularizers import l2
+from keras.optimizers import RMSprop, SGD
+from sklearn.cross_validation import train_test_split
+import tensorflow as tf
```
Comment (snf) [inline examples/resnet_cifar10.py]: I think you forgot to remove the tensorflow import. This looks really good.

Comment (pengpaiSH) [inline examples/resnet_cifar10.py]: @snf A ha. Thank you for your feedback!

File: examples/resnet_cifar10.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    nb_epoch = 25
+    data_augment = True
+
+    #  X_train, X_test: (nb_samples, 3, 32, 32)
+    (X_train, y_train), (X_test, y_test) = cifar10.load_data()
+    X_train = X_train.astype('float32')
+    X_test = X_test.astype('float32')
+    # perpixel mean substracted
+    X_train = (X_train - np.mean(X_train))/np.std(X_train)
+    X_test = (X_test - np.mean(X_test))/np.std(X_test)
```
Comment (giorgiop) [inline examples/resnet_cifar10.py]: You should use the same transformation for `X_test` here

File: examples/resnet_cifar10.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    model.add(Convolution2D(16, 3, 3, subsample = (1, 1), border_mode='same', input_shape=(3, 32, 32)))
+
+    # 1st 2n layer, 16 filters, output shape: (16, 32, 32)
+    for i in range(2 * n):
+        model.add(building_residual_block(input_shape = (16, 32, 32), n_feature_maps = 16,
+                  kernel_sizes = (3, 3), n_skip = 2, is_subsample = False, subsample = None))
+
+    # 2nd 2n layer, 32 filters, output shape: (32, 16, 16)
+    for i in range(2 * n):
+        # expand dimensions and half the outpu shape size
```
Comment (giorgiop) [inline examples/resnet_cifar10.py]: output

File: examples/resnet_cifar10.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    if is_subsample:
+        print ('      - with subsample:', subsample)
+    kernel_row, kernel_col = kernel_sizes
+    # set input
+    x = Input(shape=(input_shape))
+    # ***** SHORTCUT PATH *****
+    if is_subsample: # subsample (+ channel expansion if needed)
+        shortcut_y = Convolution2D(n_feature_maps, kernel_row, kernel_col,
+                                   subsample=subsample, W_regularizer=l2(0.0001),
+                                   border_mode='valid')(x)
```
Comment (giorgiop) [inline examples/resnet_cifar10.py]: I think you should use `init='he_uniform'` or `init='he_uniform'` (not sure which one)

File: examples/resnet_cifar10.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+    best_model = ModelCheckpoint(best_model_file, verbose = 1, save_best_only = True)
+
+    # this will do preprocessing and realtime data augmentation
+    datagen = ImageDataGenerator(
+        featurewise_center = False,  # set input mean to 0 over the dataset
+        samplewise_center = False,  # set each sample mean to 0
+        featurewise_std_normalization = False,  # divide inputs by std of the dataset
+        samplewise_std_normalization = False,  # divide each input by its std
+        zca_whitening = False,  # apply ZCA whitening
+        rotation_range = 10,  # randomly rotate images in the range (degrees, 0 to 180)
```
Comment (giorgiop) [inline examples/resnet_cifar10.py]: In the original experiments no rotation was used

File: examples/resnet_cifar10.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+    batch_size = 128
+    nb_epoch = 25
+    data_augment = True
+
+    #  X_train, X_test: (nb_samples, 3, 32, 32)
+    (X_train, y_train), (X_test, y_test) = cifar10.load_data()
+    X_train = X_train.astype('float32')
+    X_test = X_test.astype('float32')
+    # perpixel mean substracted
+    X_train = (X_train - np.mean(X_train))/np.std(X_train)
```
Comment (giorgiop) [inline examples/resnet_cifar10.py]: Just to be sure: does this perform a "per-pixel" mean?

File: examples/resnet_cifar10.py [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+        # expand dimensions and half the outpu shape size
+        if i == 0:
+            model.add(building_residual_block(input_shape = (32, 16, 16), n_feature_maps = 64,
+                      kernel_sizes = (3, 3), n_skip = 2, is_subsample = True, subsample = (2, 2)))
+        else:
+            model.add(building_residual_block(input_shape = (64, 8, 8), n_feature_maps = 64,
+                      kernel_sizes = (3, 3), n_skip = 2, is_subsample = False, subsample = None))
+
+    # global averaging layer
+    model.add(AveragePooling2D(pool_size = (7, 7)))
```
Comment (giorgiop) [inline examples/resnet_cifar10.py]: (7, 7) or (8,8) ?

File: examples/resnet_cifar10.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+    [1] Building block of layers for residual learning.
+        Code based on https://github.com/ndronen/modeling/blob/master/modeling/residual.py
+        , but modification of (perhaps) incorrect relu(f)+x thing and it's for conv layer
+    [2] MaxPooling is used instead of strided convolution to make it easier
+        to set size(output of short-cut) == size(output of conv-layers).
+        If you want to remove MaxPooling,
+           i) change (border_mode in Convolution2D in shortcut), 'same'-->'valid'
+           ii) uncomment ZeroPadding2D in conv layers.
+               (Then the following Conv2D is not the first layer of this container anymore,
+                so you can remove the input_shape in the line 101, the line with comment #'OPTION' )
```
Comment (giorgiop) [inline examples/resnet_cifar10.py]: Where is ZeroPadding to uncomment?

**Comment by giorgiop** [comment]
Looks good! 
I also suggest to re-write it with the functional syntax of 1.0 if you have time

**Comment by pengpaiSH** [comment]
@giorgiop Thank you for your feedback! By the way, what do you exactly mean "with functional syntax"?

**Comment by giorgiop** [comment]
Have a look [here](http://blog.keras.io/introducing-keras-10.html).

File: examples/resnet_cifar10.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+    '''
+        The 6n+2 network architecture for CIFAR-10 in [1].
+    '''
+    model = Sequential() # it's a CONTAINER, not MODEL
+
+    # 1st layer is a 3x3 conv, output shape: (16, 32, 32)
+    model.add(Convolution2D(16, 3, 3, subsample = (1, 1), border_mode='same', input_shape=(3, 32, 32)))
+
+    # 1st 2n layer, 16 filters, output shape: (16, 32, 32)
+    for i in range(2 * n):
```
Comment (giorgiop) [inline examples/resnet_cifar10.py]: Do you need `2 * n` if you call the building function with `n_skip=2` ?

File: examples/resnet_cifar10.py [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+    if is_subsample: # subsample (+ channel expansion if needed)
+        shortcut_y = Convolution2D(n_feature_maps, kernel_row, kernel_col,
+                                   subsample=subsample, W_regularizer=l2(0.0001),
+                                   border_mode='valid')(x)
+    else: # channel expansion only (e.g. the very first layer of the whole networks)
+        if is_expand_channels:
+            shortcut_y = Convolution2D(n_feature_maps, 1, 1,
+                                       W_regularizer=l2(0.0001), border_mode='same')(x)
+        else:
+            # if no subsample and no channel expension, there's nothing to add on the shortcut.
```
Comment (giorgiop) [inline examples/resnet_cifar10.py]: typo: expansion

**Comment by xuhuaren** [comment]
How can I download them, thanks. Actually, I wrote down a 3D version and wanna compare with others.

**Comment by pengpaiSH** [comment]
@xuhuaren You can find these two .py files here
https://github.com/pengpaiSH/keras/tree/master/examples

**Comment by xuhuaren** [comment]
@pengpaiSH Thanks a lot! residual net and FCN were too hot :)

File: examples/resnet_cifar10.py (hunk 2 above)
Comment (ogrisel) [inline examples/resnet_cifar10.py]: As you already have a dependency on sklearn you can use:

```python3
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)
```

Comment (pengpaiSH) [inline examples/resnet_cifar10.py]: @ogrisel Thank you and it looks simple : )

File: examples/resnet_cifar10.py [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+from sklearn.cross_validation import train_test_split
+import tensorflow as tf
+
+
+# Allocate full of GPU resource to me
+# gpu_options = tf.GPUOptions(per_process_gpu_memory_fraction=1.)
+# sess = tf.Session(config=tf.ConfigProto(gpu_options=gpu_options))
+
+
+def building_residual_block(input_shape, n_feature_maps, kernel_sizes=None, n_skip=2, is_subsample=False, subsample=None):
```
Comment (ogrisel) [inline examples/resnet_cifar10.py]: This function name sounds weird to me. `build_residual_block` or just `residual_block` would feel more natural.

Comment (pengpaiSH) [inline examples/resnet_cifar10.py]: @ogrisel Aha, you are right!

**Comment by ogrisel** [comment]
Also you might want to add an option to widen the residual block layers (more channels per convolution) as done in https://github.com/szagoruyko/wide-residual-networks . This should make it possible to train smaller networks (for n = 2 or n = 4) that are competitive with deepers resnets while better using the GPU and therefore training faster.

File: examples/resnet_cifar10.py (hunk 7 above)
Comment (esube) [inline examples/resnet_cifar10.py]: The average Pooling need to be with stride size of (1,1), the default will use a stride from the pooling. Also, you will need 8,8 pooling not 7,7 that is for imagenet.

File: examples/resnet_cifar10.py [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+                      kernel_sizes = (3, 3), n_skip = 2, is_subsample = True, subsample = (2, 2)))
+        else:
+            model.add(building_residual_block(input_shape = (64, 8, 8), n_feature_maps = 64,
+                      kernel_sizes = (3, 3), n_skip = 2, is_subsample = False, subsample = None))
+
+    # global averaging layer
+    model.add(AveragePooling2D(pool_size = (7, 7)))
+
+    model.add(BatchNormalization(axis=1))
+    model.add(Activation('relu'))
```
Comment (esube) [inline examples/resnet_cifar10.py]: I don't think you need BN and activation at this stage although it might not hurt, it is not compatible with the original implementation.

File: examples/resnet_cifar10.py (hunk 4 above)
Comment (ppwwyyxx) [inline examples/resnet_cifar10.py]: The original experiments use normal distribution.

**Comment by anewlearner** [comment]
@pengpaiSH
Thanks for sharing your code.
I have some questions about the code. Can you help me?
In your function 

> residual_model(n)

you said that, a `20(6*n+2)` layers network can be created by `n=3`. The layers here mean a single layer such as conv layer, not a residual block which includes two conv layers and a shortcut, you can refer to the 34-layer residual network in the original paper.
If my understanding is correct, `12(2*n*2)` layers will be created by the following code. So it won't be 20 layers in total when `n=3`.
```
for i in range(2 * n):
    model.add(building_residual_block(input_shape = (16, 32, 32), n_feature_maps = 16,
                                      kernel_sizes = (3, 3), n_skip = 2, is_subsample = False, subsample = None))
```
